module/zfs/dmu.c

```diff
@@ -966,6 +1003,188 @@ dmu_free_range(objset_t *os, uint64_t object, uint64_t offset,
+	*buf = abd_get_offset_size(data, buf_off, buf_size);
+
+	if (pre_buf || post_buf) {
+		*mbuf = abd_alloc_gang_abd();
+		if (pre_buf)
+			abd_gang_add(*mbuf, pre_buf, B_TRUE);
+		abd_gang_add(*mbuf, *buf, B_TRUE);
+		if (post_buf)
+			abd_gang_add(*mbuf, post_buf, B_TRUE);
+	} else {
+		*mbuf = *buf;
+	}
+}
+
+static void
+dmu_read_abd_done(zio_t *zio)
+{
+	if (abd_is_gang(zio->io_abd))
+		abd_free(zio->io_abd);
+	else
+		abd_put(zio->io_abd);
+}
+
+static int
+dmu_read_abd(dnode_t *dn, uint64_t offset, uint64_t size,
+    abd_t *data, uint32_t flags)
+{
+	spa_t *spa = dn->dn_objset->os_spa;
+	dmu_buf_t **dbp;
+	int numbufs, err;
+	zio_t *rio;
+
+	ASSERT(flags & DMU_DIRECTIO);
+
+	/*
+	 * Direct IO must be page aligned
+	 */
+	ASSERT(IO_PAGE_ALIGNED(offset, size));
+
+	err = dmu_buf_hold_array_by_dnode(dn, offset,
+	    size, B_FALSE, FTAG, &numbufs, &dbp, 0);
+	if (err)
+		return (err);
+
+	rio = zio_root(spa, NULL, NULL, ZIO_FLAG_CANFAIL);
+
+	for (int i = 0; i < numbufs; i++) {
+		dmu_buf_impl_t *db = (dmu_buf_impl_t *)dbp[i];
+		abd_t *buf = NULL, *mbuf;
+		zio_t *zio;
+
+		mutex_enter(&db->db_mtx);
+		blkptr_t *bp = dmu_buf_get_bp(db);
+
+		/* no need to read if hole or data is cached */
+		if (bp == NULL || BP_IS_HOLE(bp) || db->db_state == DB_CACHED) {
+			size_t aoff = offset < db->db.db_offset ?
+			    db->db.db_offset - offset : 0;
+			size_t boff = offset > db->db.db_offset ?
+			    offset - db->db.db_offset : 0;
+			size_t len = MIN(size - aoff, db->db.db_size - boff);
+			if (db->db_state == DB_CACHED)
+				abd_copy_from_buf_off(data,
+				    (char *)db->db.db_data + boff, aoff, len);
+			else
+				abd_zero_off(data, aoff, len);
+			mutex_exit(&db->db_mtx);
+			continue;
+		}
+
+		make_abd_for_dbuf(db, data, offset, size, &buf, &mbuf);
+		mutex_exit(&db->db_mtx);
+
+		zio = zio_read(rio, spa, bp, mbuf, db->db.db_size,
+		    dmu_read_abd_done, NULL,
+		    ZIO_PRIORITY_SYNC_READ, 0, NULL);
+
+		if (i+1 == numbufs)
+			err = zio_wait(zio);
+		else
+			zio_nowait(zio);
+	}
+
+	if (err)
+		(void) zio_wait(rio);
+	else
+		err = zio_wait(rio);
+
+	dmu_buf_rele_array(dbp, numbufs, FTAG);
+
+	return (err);
+}
+
+/*
+ * Checks whether the requested Direct IO operation is valid. This function
+ * returns either:
+ * 1		- valid Direct IO requeset
+ * 0		- invalid Direct IO request, but can be bypassed
+ * EINVAL	- invalid IO request due to alignment
+ */
+int
+dmu_check_direct_valid(dnode_t *dn, uint64_t offset, uint64_t size,
+    boolean_t write)
+{
+	objset_t *obj;
+	ASSERT3P(dn, !=, NULL);
+	int ret = 1;
+
+	obj = dn->dn_objset;
+
+	if (DMU_OS_DIRECT_IS_DISABLED(obj)) {
+		/*
+		 * In the case of disabled direct we simply just pass
+		 * the IO op off to the ARC.
+		 */
+		ret = 0;
+	} else if (DMU_OS_DIRECT_IS_ENABLED(obj)) {
+		/*
+		 * If the direct property is set to default or always the IO
+		 * request at a minimum must be PAGE_SIZE aligned.
+		 */
+		if (!IO_PAGE_ALIGNED(offset, size))
+			return (EINVAL);
```

Comment:
```suggestion
			return (SET_ERROR(EINVAL));
```

nit: our convention is place the `SET_ERROR` where the constant errno is returned.  So it should be used here and not in `dmu_write_uio_dnode()`.

Reply:
We fixed up all the return value stuff in the latest push.